### ISSUE TITLE
executor::test::test_ping_pong_yield is flaky

### ISSUE THREAD

**Issue by matklad** [DataDog/scipio#120]
This test fails for me once in a while:

```
$ cargo t -q --lib
running 95 tests
..................F............................................................................
failures:

---- executor::test::test_ping_pong_yield stdout ----
thread 'executor::test::test_ping_pong_yield' panicked at 'assertion failed: counter.get() < 10', scipio/src/executor.rs:1609:9

```

**Comment by glommer** [comment]
What was the value of the counter when it failed ?
I'll take a look early tomorrow.

**Comment by matklad** [comment]
Didn’t printed it, will look into this tomorrow. Seems like he assert! can use a message :)

**Comment by glommer** [comment]
Actually reading the code again to refresh my memory, that assertion cannot ever be anything other than 10, as it asserts right when it fires. 

What it would be more useful then is to just remove the assert and print the values of "me" as the queue is selected.

**Comment by glommer** [comment]
all right, I just managed to reproduce this by using taskset to make all tests run in the same CPU.
I have 12 CPUs so by default the load of all tests will be spread out a bit more.

Scheduler tests are quite sensitive to load, so that's likely why. I'll see how I can make this a bit more robust while still testing that we change queues often enough...

**Comment by glommer** [comment]
Ok, it seems to me like under load it just may happen that we'll schedule the same task more often

```
me 0 cnt 9
me 1 cnt 1
me 1 cnt 2
me 1 cnt 3
me 1 cnt 4
me 1 cnt 5
me 1 cnt 6
me 1 cnt 7
me 1 cnt 8
me 1 cnt 9
me 1 cnt 10
me 1 cnt 11
me 1 cnt 12
me 1 cnt 13
me 0 cnt 1
```

**Comment by glommer** [comment]
I believe the reason for that is that the quanta of work is much smaller than the OS time slice.
So what happens is that one of our task queues is asked to run for 200us, but the OS may take us out for a much longer time to run something else. The more threads you have (Which will increase with more tests), the more likely it is that you will not run for a much longer time.

When you are back to the CPU, let's say some 10ms later, the scheduler can't differentiate between your task queue having naturally run for 10ms or being out of the CPU for all that time and accounts 10ms for you. You will now be penalized for a while...

Making the scheduler more accurate is not worth it because it is too expensive, and good TPC applications will and should just pin.

For the purposes of making this test more robust I think we should just increase the work quanta. The same phenomenon will still happen but if you are accounted 3ms for a 2ms run, that's much different than being accounted 2ms for a 200us run. 

Note that if we could make this perfectly robust and always pinned, we wouldn't need the counter to 10. It should ping pong perfectly between queues. The broken scheduler that motivated this test, on the other hand, would run all task queues until the preempt timer hit forcing them out (100ms), even if they had yielded.

**Comment by glommer** [comment]
It would be really nice to be able to say that some tests should run alone, btw.
I found some crates to do serial testing, but that's not really what we need: they'd require marking pretty much all tests are serial. What we need is just a handful of executor tests to execute undisturbed and the rest can go wild.

But trying to make them more robust is a good exercise too =)

**Comment by matklad** [comment]
>What we need is just a handful of executor tests to execute undisturbed
and the rest can go wild.

we can lock a reader-write lock at the start of the test. undisturbable
tests would w, and the rest would r.

**Comment by glommer** [comment]
Right, but that would require every test that we always write to take that lock, no ?
I thought of that but sounded quite heavy-handed.

**Comment by matklad** [comment]
It surely is, but almost every test needs to create an executor anyway, so
we already have some common-to-all tests boilerplate, so I don't think it
would be more annoying in practice than what we have today. Locking can be
hidden in the already existing macros.

But, surely, making the tests more robust to threading would be a cleaner
option :0)
